### PR TITLE
base experiments and environments

### LINKED ISSUE #2: JaxEnvironment
Derived from `artiq.experiments.HasEnvironment`. This class should provide basic functionalities of an experiment, including the followings:
* Connect to labrad
* Data saving and loading shared data
* Parameter loading and converting to an [ARTIQ python readable type](https://m-labs.hk/artiq/manual-beta/compiler.html#supported-python-features)
* With a class variable `USE_LABRAD`, inherited classes can disable automatic connection to labrad, which is used in the case where labrad is not running on the control computer (or a labrad connection is not needed). This should only apply to the simplest experiments with no data saving / parameter loading requirements. I am not sure what is the use case here, but it is easy to implement anyways.

### LINKED ISSUE #3: JaxExperiment and SinaraExperiment
`JaxExperiment ` inherits from `artiq.experiment.Experiment`. Base class for all labwide experiments. In this class, we should implement common things for lab experiments that don't depend on the sinara hardware. I cannot think of anything we should implement in this class right now, so it might just be an inheritance with no modification. We still want this class for possibly adding things here in the future.

`SinaraExperiment` inherits from `JaxExperiment`. It implements common things for all experiments that use the sinara hardware, including getting devices, resetting devices, keeping track of DDS/TTL values in a scalable manner.

### LINKED ISSUE #5: pulse sequence class
Base class for all pulse sequences. Provides scalable management of required parameters, required DDS/TTL channels (so they can be initialized correctly in the experiment).

### LINKED ISSUE #6: device manager
A class that can be used in ARTIQ python that keeps track of DDS/TTLs used by an experiment.

For DDSes, they all need to be turned off at the start of the experiment, and reset to previous values at the end.
For TTLes, they may need to be set to the correct polarity (input/output) at the start of the experiment, and reset to previous values at the end.

### PR DESCRIPTION
**Pull request by fanmingyu212**
Closes #2. Closes #3. Closes #5. Closes #6.

An actual experiment class needs to inherit from a base experiment and an environment.

Base experiments:
* `JaxExperiment`: Just wrapper over the base `artiq.experiment.Experiment` class.
* We need scanning experiments (#4) that are based on `JaxExperiment`

Environments:
* `JaxEnvironment`: Environment for all experiments. If the experiment does not use sinara, it should inherit from this environment. Provides functions to handle data, parameters, and drift trackers.
* `SinaraEnvironment`: Environment for all experiments dependent on the sinara hardware. Provides functions to do common operations with sinara kernel code, e.g., setting DDS/TTL values (useful for resetting parameters at the end of an experiment). 

Common experiments that are shared by different setups:
* `InitializeSinara`: Initializes DDS/TTL of a sinara box. Sets the parameters if the devices are known, and saves the parameters if they are unknown.
* `PMT`: Background experiment that should run all the time. Gets PMT counts and enables setting DDS/TTL parameters in real time (~0.1 s delay).

Example experiments:
* `HardwareControl`: How to write a simple experiment and do basic hardware control.
* `PulseSequenceAndDataSaving`: How to use a pulse sequence, parameters, drift trackers, and how to save data.

Sequences:
* `Sequence`: Base pulse sequence class.

Example pulse sequences:
* `DopplerCool`
* `StateDetect`
* `Example2`: example pulse sequence that is used in the `PulseSequenceAndDataSaving` experiment.

Utilties:
* `Devices`: Provides lists of different types of devices from "devices_db.py" (device definition file). Also keeps track of devices used in an experiment for initialization and resetting.
* `labrad.py`: Provides a function to convert `WithUnit` arrays to `numpy.array`.
* `TreeDict`: Provides a base class for building ARTIQ python compatible object saving values in a dictionary. See `ParameterGroup` below.
* `ParameterGroup`: Inherits from `TreeDict`. Stores parameters so they can be accessed in ARTIQ python.
* `DriftTracker`: Inherits from `TreeDict`. Provides conversion from detuning to DDS frequency. Compatible with ARTIQ python.

### CODE REVIEW

**Comment by fanmingyu212** [comment]
I will make some improvements on names, docstrings, and accompanying wiki pages. Then I will convert this back to a PR.

File: experiments/common/initialize_sinara.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    This experiment assumes that the sinara hardware has at least a DDS and a TTL board.
+    For an uninitialized DDS or TTL device, the experiment assumes its state to be off (low).
+    """
+    kernel_invariants = {
+        "initialized_ddses", "uninitialized_ddses", "initialized_ttls", "uninitialized_ttls"
+    }
+
+    def build(self):
+        super().build()
+        second_lowest_priority = -99  # we use priorities range from -100 to 100.
```
Comment (mczimmermann) [inline experiments/common/initialize_sinara.py:27]: why does it initialize to the second lowest priority?

File: experiments/common/initialize_sinara.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            else:
+                self.cxn.artiq.remove_sinara_ttl(name)
+        for name in self.devices.ttl_outs:
+            if name not in ttl_params:
+                self.uninitialized_ttls.append((self.get_device(name), name))
+
+    @kernel
+    def run_kernel(self):
+        self.core.reset()
+        for kk in range(len(self.initialized_ddses)):
```
Comment (mczimmermann) [inline experiments/common/initialize_sinara.py:77]: what does kk stand for

File: experiments/common/initialize_sinara.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        """Get the DDS values from the sinara hardware.
+
+        It gets the frequency, phase, amplitude, and attenuation from the DDS,
+        and assumes that the DDS state is -1. (off).
+        """
+        self.core.break_realtime()
+        frequency, phase, amplitude = device.get()
+        self.core.break_realtime()
+        attenuation = device.get_att()
+        state = -1.
```
Comment (mczimmermann) [inline experiments/common/initialize_sinara.py:108]: for our experiments is it more efficient to assume that the DDS state is off/if it's on will this create errors elsewhere?

File: experiments/common/initialize_sinara.py (hunk 2 above)
Comment (fanmingyu212) [reply experiments/common/initialize_sinara.py:77]: `kk` is just an integer counting the index when we loop through the `self.initialized_ddses` in the for loop. See https://www.askpython.com/python/list/iterate-through-list-in-python (first method) for example.

**Pull request merged**